DREAM/Optimization/tsgParticleSwarm.cpp

```diff
@@ -0,0 +1,146 @@
+
+    // Initialize helper variables and functions.
+    size_t num_dimensions = (size_t) state.getNumDimensions();
+    size_t num_particles = (size_t) state.getNumParticles();
+    std::vector<double> &particle_positions = state.getParticlePositionsRef();
+    std::vector<double> &particle_velocities = state.getParticleVelocitiesRef();
+    std::vector<double> &best_particle_positions = state.getBestParticlePositionsRef();
+    std::vector<double> &best_position = state.getBestPositionRef();
+    std::vector<double> prev_best_vals(num_particles), cur_best_vals(num_particles);
+    std::vector<bool> prev_is_inside(num_particles), cur_is_inside(num_particles);
+    double best_fval = std::numeric_limits<double>::max();
+    ObjectiveFunctionConstrained f_constrained = makeObjectiveFunctionConstrained(num_dimensions, f, inside);
+
+    // Create a lambda that updates the best particle and swarm positions based on function values and domain information.
+    auto update_best_positions = [&](const std::vector<double> prev_vals, const std::vector<double> cur_vals,
+                                     const std::vector<bool> prev_domain, const std::vector<bool> cur_domain)->void {
+        for (size_t i=0; i<num_particles; i++) {
+            bool is_smaller = (cur_vals[i] < prev_vals[i]) || !prev_domain[i];
+            if (cur_domain[i] && is_smaller) {
+                std::copy(particle_positions.begin() + i * num_dimensions, particle_positions.begin() + (i + 1) * num_dimensions,
+                          best_particle_positions.begin() + i * num_dimensions);
+                if (cur_vals[i] < best_fval) {
+                    std::copy(particle_positions.begin() + i * num_dimensions, particle_positions.begin() + (i + 1) * num_dimensions,
+                              best_position.begin());
+                    best_fval = cur_vals[i];
+                }
+            }
+        }
+    };
+
+    // Initialize the best particle and swarm function values, while updating the input state.
+    f_constrained(best_particle_positions, prev_best_vals, prev_is_inside);
+    f_constrained(particle_positions, cur_best_vals, cur_is_inside);
+    update_best_positions(prev_best_vals, cur_best_vals, prev_is_inside, cur_is_inside);
+    prev_best_vals = cur_best_vals;
+    prev_is_inside = cur_is_inside;
+
+    // Main algorithm starts here.
+    while(state.getNumIterations() < max_iterations) {
+        for (size_t i=0; i<num_particles; i++) {
+            for (size_t j=0; j<num_dimensions; j++) {
+                double rp = get_random01();
+                double rg = get_random01();
+                int idx = i * num_dimensions + j;
+                particle_velocities[idx] = inertia_weight * particle_velocities[idx] +
+                                           cognitive_coeff * rp * (best_particle_positions[idx] - particle_positions[idx]) +
+                                           social_coeff * rg * (best_position[j] - particle_positions[idx]);
```

Comment:
So you have `i` and `j` as `size_t` which is 64-bit, but when you multiply them and get a much larger number, you have `idx` as an `int`.

Instead of having double-index logic, note that only `best_position` uses a single index, so you can put all of this in a single loop:
```
for(size_t i=0; i<num_particles * num_dimensions; i++){
...
            social_coeff * rg * (best_position[i % num_dimensions] - particle_positions[i]);
...
}
```

You can also use the `Utils::size_mult()` template in tsgUtils.hpp that will convert both `num_particles` and `num_dimensions` to `size_t` and avoid annoying warnings.

On a side-note, are you using `-Wall -Wextra -Wshadow` for your internal testing. Sometimes those are annoying, but sometimes they help catch bugs.

Reply:
Changed to only iterating over `i`.

Reply:
For my testing: yes, my flags from CMake are `CMAKE_CXX_FLAGS="-O3 -Wall -Wextra -Wshadow"`.